### PR TITLE
Generate and validate with json schema for polymer.json

### PR DESCRIPTION
**Pull request by rictic**

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -11,10 +11,15 @@
 import * as fs from 'fs';
 import * as path from 'path';
 import * as logging from 'plylog';
+import * as jsonschema from 'jsonschema';
+
 import minimatchAll = require('minimatch-all');
 
 const logger = logging.getLogger('polymer-project-config');
 
+const schema = JSON.parse(
```
Comment (justinfagnani) [inline src/index.ts]: Let's try to not do sync file IO during module initialization. Add a `getSchema()` function that can lazily load it. It'd be even better if we avoided sync IO.

File: src/index.ts [hunk 2]
```diff
@@ -134,10 +139,19 @@ export class ProjectConfig {
   static loadOptionsFromFile(filepath: string): ProjectOptions {
     try {
       const configContent = fs.readFileSync(filepath, 'utf-8');
-      return JSON.parse(configContent);
+      const contents = JSON.parse(configContent);
```
Comment (justinfagnani) [inline src/index.ts]: While you're here, can you add a TODO to make this method async?

Comment (rictic) [inline src/index.ts]: Done

File: src/index.ts (hunk 1 above)
Comment (rictic) [inline src/index.ts]: Fair enough. Done, with TODO to become async.

File: src/index.ts [hunk 3]
```diff
@@ -130,14 +132,25 @@ export class ProjectConfig {
...
+        throw result.throwError;
+      }
+      if (result.errors.length > 0) {
+        throw result.errors[0];
+      }
+      return contents;
     } catch (error) {
       // swallow "not found" errors because they are so common / expected
-      if (error.code === 'ENOENT') {
+      if (error && error.code === 'ENOENT') {
```
Comment (justinfagnani) [inline src/index.ts]: was someone throwing `null`?

File: src/index.ts [hunk 4]
```diff
@@ -315,3 +328,18 @@ export class ProjectConfig {
   }
 
 }
+
+// Gets the json schema for polymer.json, generated from the typescript
+// interface for runtime validation. See the build script in package.json for
+// more info.
+const getSchema: () => jsonschema.Schema = (() => {
```
Comment (justinfagnani) [inline src/index.ts]: this is a little complex. Why not just:

```typescript
let schema: jsonschema.Schema | undefined = undefined;
function getSchema() {
  if (schema === undefined) {
    schema = JSON.parse(
        fs.readFileSync(path.join(__dirname, 'schema.json'), 'utf-8'));
  }
  return schema;
}
```

you're only hiding the variable from ourselves with the closure.

File: src/index.ts (hunk 3 above)
Comment (rictic) [inline src/index.ts]: No, I just figured it was better to check here than not to

File: src/index.ts (hunk 4 above)
Comment (rictic) [inline src/index.ts]: This way is more obviously correct. Yeah we're only hiding it from ourselves, but this way you don't have to think about the bug of accessing a potentially uninitialized variable.

**Pull request merged**